en/guide/message_signing.md

```diff
@@ -0,0 +1,121 @@
+```
+
+The maximum number of signing streams supported is given by the `MAVLINK_MAX_SIGNING_STREAMS` macro. This defaults to 16, but it may be worth raising this for GCS implementations. If the C implementation runs out of signing streams then new streams will be rejected.
+
+## Using the accept_unsigned_callback
+
+In the signing structure there is an optional `accept_unsigned_callback` function pointer. The C prototype for this function is:
+
+```
+bool accept_unsigned_callback(const mavlink_status_t *status, uint32_t msgId);
+```
+
+If set in the signing structure then this function will be called on any unsigned packet (including all *MAVLink 1* packets) or any packet where the signature is incorrect. The function offers a way for the implementation to allow unsigned packets to be accepted.
+
+The rules for what unsigned packets should be accepted is implementation specific, but it is suggested the following rules be considered:
+
+* have a mechanism for marking a particular communication channel as being secure (such as a USB connection) to allow for signing setup.
+* always accept `RADIO_STATUS` packets for feedback from 3DR radios (which don't do signing)
+
+For example:
+
+```c
+static const uint32_t unsigned_messages[] = {
+	MAVLINK_MSG_ID_RADIO_STATUS
+};
+
+static bool accept_unsigned_callback(const mavlink_status_t *status, uint32_t message_id)
+{
+	// Make the assumption that channel 0 is USB and should always be accessible
+	if (status == mavlink_get_channel_status(MAVLINK_COMM_0)) {
+		return true;
+	}
+
+	for (unsigned i = 0; i < sizeof(unsigned_messages) / sizeof(unsigned_messages[0]); i++) {
+		if (unsigned_messages[i] == message_id) {
+			return true;
+		}
+	}
+
+	return false;
+}
+```
+
+## Handling Link IDs
+
+The purpose of the `link_id` field in the *MAVLink 2* signing structure is to prevent cross-channel replay attacks. Without the `link_id` an attacker could record a packet (such as a disarm request) on one channel, then play it back on a different channel.
+
+The intention with the link IDs is that each channel of communication between an autopilot and a GCS uses a different link ID. There is no requirement that the same link ID be used in both directions however.
+
+For C implementations the obvious mechanism is to use the MAVLink channel number as the link ID. That works well for an autopilot, but runs into an issue for a GCS implementation. The issue is that a user may launch multiple GCS instances talking to the same autopilot via different communication links (such as two radios, or USB and a radio). These multiple GCS instances will not be aware of each other, and so may choose the same link ID. If that happens then a large number of correctly signed packets will be rejected by the autopilot as they will have timestamps that are older than the timestamp received for the same stream tuple on the other communication link.
+
+The solution adopted for MAVProxy is shown below:
+
+```
+if (msg.get_signed() and
```

Comment:
Sorry, only saw this now - we need to replace this with an example from one of the maintained codebases. Or better: Generalize this to general MAVLink C++ code so we have all examples consistently in C++.

Reply:
Can you point me to anything that uses this - quick search on `link_id`  shows no usage in the mavlink projects. 

Bigger deal is that I don't understand how this example can work. According to the text the problem is that "These multiple GCS instances will not be aware of each other, and so may choose the same link ID." However if your solution is to set the GCS ID from the autopilot link ID, and it is choosing its ID from the MAVLink channel number, then you're guaranteeing that all GCS connections to that autopilot WILL share the same id.